### PR TITLE
Sequential resource processing

### PR DESCRIPTION
**Pull request by abdermendz**
Description:
This Pull Request contains changes for introducing new parameter 'processed-after'.
With this parameter sequential resource processing is now possible.

Backlog Item: LMCROSSITXSADEPLOY-1874

### CODE REVIEW

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/variables/Variables.java [hunk 1]
```diff
@@ -589,4 +593,16 @@ public Duration deserialize(Object serializedObject) {
                                                                                        .defaultValue(true)
                                                                                        .build();
 
+    ImmutableJsonStringListVariable<List<Resource>> BATCHES_TO_PROCESS = ImmutableJsonStringListVariable.<List<Resource>> builder()
```
Comment (radito3) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/variables/Variables.java:596]: Use just the interface here -> `Variable<List<Resource>>`
And shouldn't this be `Variable<List<List<Resource>>>`?

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/BuildCloudDeployModelStep.java [hunk 2]
```diff
@@ -112,22 +112,28 @@ protected StepPhase executeStep(ProcessContext context) {
...
-        // Needed by CreateOrUpdateServicesStep, as it is used as an iteration variable:
-        context.setVariable(Variables.SERVICES_TO_CREATE_COUNT, servicesToCreate.size());
+        // logg batching of resources
+        getStepLogger().debug(Messages.CALCULATING_RESOURCE_BATCHES);
+        // Organise resources into batches of resources
+        List<List<Resource>> batchesToProcess = calculateBatchesToProcess(deploymentDescriptor, resourcesForDeployment);
+        // set context variable batches_to_process
+        context.setVariable(Variables.BATCHES_TO_PROCESS, batchesToProcess);
+        // log end of the batching
+        getStepLogger().debug(Messages.CALCULATING_RESOURCE_BATCHES_COMPLETE);
```
Comment (IvanBorislavovDimitrov) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/BuildCloudDeployModelStep.java:122]: These comments are reduntant

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/BuildCloudDeployModelStep.java [hunk 3]
```diff
@@ -37,6 +36,7 @@
 import org.cloudfoundry.multiapps.mta.builders.v2.ParametersChainBuilder;
 import org.cloudfoundry.multiapps.mta.handlers.HandlerFactory;
 import org.cloudfoundry.multiapps.mta.handlers.v2.DescriptorHandler;
+import org.cloudfoundry.multiapps.mta.handlers.v3.BatchCalculator;
```
Comment (IvanBorislavovDimitrov) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/BuildCloudDeployModelStep.java:39]: What will happen if processed after is used with schema_version 2?

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/ExtractServicesDeploymentStep.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Override
+    protected StepPhase executeStep(ProcessContext context) {
+        getStepLogger().debug(Messages.EXTRACT_SERVICES_FROM_BATCH);
+        // get singular batch of resources variable
+        List<Resource> batchOfResources = context.getVariable(Variables.BATCH_TO_PROCESS);
+        // extract all services from the batch of resources
+        ServicesCloudModelBuilder servicesCloudModelBuilder = getServicesCloudModelBuilder(context);
+        List<CloudServiceInstanceExtended> servicesCalculatedForDeployment = servicesCloudModelBuilder.build(batchOfResources);
+        // Build a list of services for creation and save them in the context:
```
Comment (IvanBorislavovDimitrov) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/ExtractServicesDeploymentStep.java:30]: I'm not a fan of comments, and I think the code describes itself enough, I'd suggest to remove them

**Review by IvanBorislavovDimitrov — COMMENTED** [review]
Have you tested, cf undeploy --delete-services?

**Pull request merged**